### PR TITLE
Implementing WithCorrelationId method

### PR DESCRIPTION
**Pull request by trwalke**
Implementation of [#1211](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/1211)

### CODE REVIEW

File: src/Microsoft.Identity.Client/ApiConfig/AbstractAcquireTokenParameterBuilder.cs [hunk 1]
```diff
@@ -328,6 +328,17 @@ public T WithB2CAuthority(string authorityUri)
             return (T)this;
         }
 
+        /// <summary>
+        /// Sets the correlation id to be used in the authentication request.
+        /// </summary>
+        /// <param name="correlationId">correlation id of the authentication request</param>
+        /// <returns></returns>
+        public T WithCorrelationId(Guid correlationId)
```
Comment (trwalke) [inline src/Microsoft.Identity.Client/ApiConfig/AbstractAcquireTokenParameterBuilder.cs]: @jmprieur @henrik-me Using GUID here for the correlation ID. Do we want to keep it as is or change it to a string input instead? The authenticationResult in this PR also contains a correlation id of type GUID

Comment (henrik-me) [reply src/Microsoft.Identity.Client/ApiConfig/AbstractAcquireTokenParameterBuilder.cs]: GUID is great

---
In reply to: [297839588](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1235#discussion_r297839588) [](ancestors = 297839588)

Comment (MarkZuber) [reply src/Microsoft.Identity.Client/ApiConfig/AbstractAcquireTokenParameterBuilder.cs]: MATS requires a guid for it's correlation ids, so my $.02 is that we make this a guid explicitly, as you have here.  That way we can format it appropriately for what the telemetry system and services are expecting (there's a AsMatsCorrelationId or something like that helper method in the .net mats telemetry code).

File: src/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenCommonParameters.cs [hunk 2]
```diff
@@ -14,11 +14,11 @@ internal class AcquireTokenCommonParameters
 
         public ApiEvent.ApiIds ApiId { get; set; } = ApiEvent.ApiIds.None;
         public Guid TelemetryCorrelationId { get; set; }
+        public Guid UserProvidedTelemetryCorrelationId { get; set; }
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenCommonParameters.cs]: >Telemetry [](start = 32, length = 9)

This is not necessarily a telemetry id... It's a CorrelationId which has been user provided. Pls. remove the word Telemetry

File: src/Microsoft.Identity.Client/Internal/Requests/ClientCredentialRequest.cs [hunk 3]
```diff
@@ -32,7 +32,7 @@ internal override async Task<AuthenticationResult> ExecuteAsync(CancellationToke
                 var msalAccessTokenItem = await CacheManager.FindAccessTokenAsync().ConfigureAwait(false);
                 if (msalAccessTokenItem != null)
                 {
-                    return new AuthenticationResult(msalAccessTokenItem, null);
+                    return new AuthenticationResult(msalAccessTokenItem, null, AuthenticationRequestParameters.RequestContext.TelemetryCorrelationId);
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/Internal/Requests/ClientCredentialRequest.cs]: is this intentional?

File: src/Microsoft.Identity.Client/Internal/Requests/OnBehalfOfRequest.cs [hunk 4]
```diff
@@ -41,7 +41,7 @@ internal override async Task<AuthenticationResult> ExecuteAsync(CancellationToke
                 MsalAccessTokenCacheItem msalAccessTokenItem = await CacheManager.FindAccessTokenAsync().ConfigureAwait(false);
                 if (msalAccessTokenItem != null)
                 {
-                    return new AuthenticationResult(msalAccessTokenItem, null);
+                    return new AuthenticationResult(msalAccessTokenItem, null, AuthenticationRequestParameters.RequestContext.TelemetryCorrelationId);
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/Internal/Requests/OnBehalfOfRequest.cs]: intentional?

File: src/Microsoft.Identity.Client/Internal/Requests/ClientCredentialRequest.cs (hunk 3 above)
Comment (trwalke) [reply src/Microsoft.Identity.Client/Internal/Requests/ClientCredentialRequest.cs]: yes. the correlation ID needs to be made available on the AuthenticationResult so I pass it into the constructor

File: src/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenCommonParameters.cs (hunk 2 above)
Comment (henrik-me) [inline src/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenCommonParameters.cs]: >UserProvidedTelemetryCorrelationId [](start = 20, length = 34)

This is the new one, seems like you have now started taking the param above and using that in a number of places. is that intentional?

File: src/Microsoft.Identity.Client/Internal/Requests/SilentRequest.cs [hunk 5]
```diff
@@ -98,7 +98,7 @@ internal override async Task<AuthenticationResult> ExecuteAsync(CancellationToke
                 if (msalAccessTokenItem != null)
                 {
                     var msalIdTokenItem = await CacheManager.GetIdTokenCacheItemAsync(msalAccessTokenItem.GetIdTokenItemKey()).ConfigureAwait(false);
-                    return new AuthenticationResult(msalAccessTokenItem, msalIdTokenItem);
+                    return new AuthenticationResult(msalAccessTokenItem, msalIdTokenItem, AuthenticationRequestParameters.RequestContext.TelemetryCorrelationId);
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/Internal/Requests/SilentRequest.cs]: >TelemetryCorrelationId [](start = 137, length = 22)

?

File: src/Microsoft.Identity.Client/ApiConfig/AbstractAcquireTokenParameterBuilder.cs [hunk 6]
```diff
@@ -340,7 +351,7 @@ internal void ValidateAndCalculateApiId()
             Validate();
             CommonParameters.ApiId = CalculateApiEventId();
             CommonParameters.ApiTelemId = ApiTelemetryId;
-            CommonParameters.TelemetryCorrelationId = Guid.NewGuid();
+            CommonParameters.TelemetryCorrelationId = (CommonParameters.UserProvidedTelemetryCorrelationId != null) ? CommonParameters.UserProvidedTelemetryCorrelationId : new Guid();
```
Comment (MarkZuber) [inline src/Microsoft.Identity.Client/ApiConfig/AbstractAcquireTokenParameterBuilder.cs]: you should use Guid.NewGuid() instead of "new Guid()".  I don't believe the Guid constructor actually calls CoCreateGuid()

File: src/Microsoft.Identity.Client/AuthenticationResult.cs [hunk 7]
```diff
@@ -43,7 +44,8 @@ public partial class AuthenticationResult
             string tenantId,
             IAccount account,
             string idToken,
-            IEnumerable<string> scopes)
+            IEnumerable<string> scopes,
+            string correlationID)
```
Comment (MarkZuber) [inline src/Microsoft.Identity.Client/AuthenticationResult.cs]: nit:  correlationId

File: tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/PublicClientApplicationTests.cs [hunk 8]
```diff
@@ -1262,6 +1262,38 @@ public void EnsurePublicApiSurfaceExistsOnInterface()
             CheckBuilderCommonMethods(byRefreshTokenBuilder);
         }
 
+        [TestMethod]
+        public void CheckUserProvidedCorrlationIDTest()
```
Comment (henrik-me) [inline tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/PublicClientApplicationTests.cs]: >Corrlation [](start = 37, length = 10)

nit: missing e

File: src/Microsoft.Identity.Client/AuthenticationResult.cs [hunk 9]
```diff
@@ -141,6 +145,11 @@ internal AuthenticationResult(MsalAccessTokenCacheItem msalAccessTokenCacheItem,
         /// </summary>
         public IEnumerable<string> Scopes { get; }
 
+        /// <summary>
+        /// Gets the correlation id used for the request.
+        /// </summary>
+        public string CorrelationID { get; }
```
Comment (MarkZuber) [inline src/Microsoft.Identity.Client/AuthenticationResult.cs]: nit:  CorrelationId  (.net naming conventions have ID as Id)

Do we want this to be TelemetryCorrelationId or is this more broad.  We're calling it TelemetryCorrelationId elsewhere.   
Should this also be a Guid type since that's how we're sending it in?

File: tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/ConfidentialClientApplicationTests.cs [hunk 10]
```diff
@@ -64,7 +64,8 @@ public void MockConfidentialClientApplication_AcquireToken()
                 {
                     "scope1",
                     "scope2"
-                });
+                },
+                new Guid().ToString());
```
Comment (MarkZuber) [inline tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/ConfidentialClientApplicationTests.cs]: Guid.NewGuid().   We should also always specify the format of the guid as we expect it.  Due to differences in how ToString() works on different platforms (some require IFormatProvider), there's a AsMatsCorrelationId() extension method in Microsoft.Identity.Client.TelemetryCore.Internal.

File: tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/PublicClientApplicationTests.cs [hunk 11]
```diff
@@ -1262,6 +1262,38 @@ public void EnsurePublicApiSurfaceExistsOnInterface()
...
         }
 
+        [TestMethod]
+        public void CheckUserProvidedCorrlationIDTest()
+        {
+            using (var harness = CreateTestHarness())
+            {
+                harness.HttpManager.AddInstanceDiscoveryMockHandler();
+                var correlationId = new Guid();
+                PublicClientApplication app = PublicClientApplicationBuilder.Create(MsalTestConstants.ClientId)
```
Comment (MarkZuber) [inline tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/PublicClientApplicationTests.cs]: nit:  please wrap these to the next line tabbed in 4 spaces to reduce width for code reviews:

PublicClientApplication app = PUblicClientApplicationBuilder
    .Create(foo)
    .WithAuthority(bar)
    .etc();

File: src/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenCommonParameters.cs (hunk 2 above)
Comment (trwalke) [reply src/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenCommonParameters.cs]: yes. This is used to set the correlation id that will be used by the authentication request.

File: src/Microsoft.Identity.Client/AuthenticationResult.cs (hunk 9 above)
Comment (trwalke) [reply src/Microsoft.Identity.Client/AuthenticationResult.cs]: Yes. it will be a Guid.
I think CorrelationId is sufficient on its own as it is a more general term.
@henrik-me @jmprieur thoughts?

Comment (henrik-me) [reply src/Microsoft.Identity.Client/AuthenticationResult.cs]: CorrelationId, is the right term imo. can be telemetry or not.

File: src/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenCommonParameters.cs (hunk 2 above)
Comment (henrik-me) [reply src/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenCommonParameters.cs]: hmm, that is not what the user passed in. that seems odd to me

File: src/Microsoft.Identity.Client/Internal/Requests/ClientCredentialRequest.cs (hunk 3 above)
Comment (henrik-me) [reply src/Microsoft.Identity.Client/Internal/Requests/ClientCredentialRequest.cs]: as mentioned elsewhere... this is the telemetrycorrelationid and not the new id. Also why are you creating this as a new Guid?

---
In reply to: [297841976](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1235#discussion_r297841976) [](ancestors = 297841976)

File: src/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenCommonParameters.cs (hunk 2 above)
Comment (trwalke) [reply src/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenCommonParameters.cs]: I dont think so but perhaps I am mistaken? I checked the code and the TelemetryCorrelationId is the same as the CorrelationId. It seems like we have two different names for the same value in the code.

File: src/Microsoft.Identity.Client/Internal/Requests/ClientCredentialRequest.cs (hunk 3 above)
Comment (trwalke) [reply src/Microsoft.Identity.Client/Internal/Requests/ClientCredentialRequest.cs]: We have two names for the same value so this is a little confusing. the telemetrycorrelationid and the correlationid are the same. Also, I create a new guid because we are exposing the correlation id as a GUID on the AuthenticationResult but the correlation id in the requestContext is exposed as a string so i have to make the conversion.

---
In reply to: [298004949](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1235#discussion_r298004949) [](ancestors = 298004949,297841976)

File: src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs [hunk 12]
```diff
@@ -98,7 +99,7 @@ private async Task AcquireAuthorizationAsync(CancellationToken cancellationToken
             AuthenticationRequestParameters.RedirectUri = _webUi.UpdateRedirectUri(AuthenticationRequestParameters.RedirectUri);
             var authorizationUri = CreateAuthorizationUri(true);
 
-            var uiEvent = new UiEvent(AuthenticationRequestParameters.RequestContext.TelemetryCorrelationId);
+            var uiEvent = new UiEvent(AuthenticationRequestParameters.RequestContext.CorrelationId.AsMatsCorrelationId());
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs]: >AsMatsCorrelationId() [](start = 99, length = 21)

why is this needed?

File: tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/ConfidentialClientApplicationTests.cs (hunk 10 above)
Comment (henrik-me) [reply tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/ConfidentialClientApplicationTests.cs]: hmm, still not sure I get this. 

---
In reply to: [297843566](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1235#discussion_r297843566) [](ancestors = 297843566)

File: tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/PublicClientApplicationTests.cs [hunk 13]
```diff
@@ -1262,6 +1263,40 @@ public void EnsurePublicApiSurfaceExistsOnInterface()
             CheckBuilderCommonMethods(byRefreshTokenBuilder);
         }
 
+        [TestMethod]
+        public void CheckUserProvidedCorrelationIDTest()
+        {
+            using (var harness = CreateTestHarness())
+            {
+                harness.HttpManager.AddInstanceDiscoveryMockHandler();
+                var correlationId = new Guid();
```
Comment (MarkZuber) [inline tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/PublicClientApplicationTests.cs]: Guid.NewGuid()

Comment (MarkZuber) [reply tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/PublicClientApplicationTests.cs]: new Guid() always returns Guid.Null (00000-00000-00000-...)

Comment (trwalke) [reply tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/PublicClientApplicationTests.cs]: Missed this one. thanks

File: src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs (hunk 12 above)
Comment (trwalke) [reply src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs]: I changed the requestContext's correlation id to a Guid as it is the main mechanism in which the correlationId moves through the project. Some classes that use the correlation Id, such as the UiEvent class require a string for the correlation Id because its base type is Dictionary<string, string> and a Guid would incompatible since it cannot be added to the collection.
